### ISSUE TITLE
Type information lost on unification

### ISSUE THREAD

**Issue by JadenGeller** [idris-lang/Idris-dev#4103]
Type information is lost when `Vect (c + (r' * c)) a` is unified with `Vect (c + n) a`. When `r' * c` unifies with `n`, the type information `n = r' * c` is thrown away.

# Steps to Reproduce
```idris
import Data.Vect
%hide Data.Vect.transpose

Matrix : Nat -> Nat -> Type -> Type
Matrix r c a = Vect r (Vect c a)

transpose : Matrix r c a -> Matrix c r a
transpose [] {c} = replicate c []
transpose ([] :: _) = []
transpose ((x :: xs) :: rows) = (x :: map head rows) :: transpose (xs :: map tail rows)

unrotate : Matrix r c a -> Matrix c r a
unrotate = reverse . transpose

spiral : Matrix r c a -> Vect (r * c) a
spiral {r=Z} [] = []
spiral {r=S r'} {c} (row :: rows) = row ++ ?bar
```

# Expected Behavior
`bar` has type `Vect (r' * c) a`. The hole can be filled with `rewrite multCommutative r' c in spiral (unrotate rows)`.

# Observed Behavior
`bar` has type `Vect n a` for all `n`. The hole cannot be filled.

The workaround is to rewrite the right hand side as `let rest : Vect (r' * c) a = rewrite multCommutative r' c in spiral (unrotate rows) in row ++ rest`.

**Comment by ahmadsalim** [comment]
@JadenGeller Thanks for reporting the issue.
Are you saying that the type of the hole is inprecise compared to what is expected?
So you get `Vect n a` instead of `Vect (r' * c) a` for `bar`.

I am not sure there is any unification going on here.

**Comment by JadenGeller** [comment]
Yes, that is what I'm saying. Sorry, I didn't use the right terminology, but you figured out what I meant.

**Comment by ahmadsalim** [comment]
@JadenGeller No need to apologize :). I was just trying to clarify what your concrete issue was.
